Add 'links' to json hash for relationships with URLs.

benkonrath opened the pull request:
I still need to write some unit tests for the URL regex so it's a work in progress.

benkonrath commented:
As I mentioned in #58, supporting this properly means users will need a custom Hyperlinked serializer field. The problem is that the `hasMany` relationship is created with multiple URLs instead of one URL that will limit only return the related items. For example:

Post
```json
{
    "id": 11, 
    "title": "post title 11", 
    "body": "post body 11", 
    "comments": [
        "http://localhost:8000/api/comments/9/", 
        "http://localhost:8000/api/comments/10/", 
        "http://localhost:8000/api/comments/11/"
    ]
}
```

It's possible to support the `belongsTo` relationship in the default DRF configuration because it uses one URL. For example:

Comment
```json
{
    "id": 9, 
    "body": "comment body 1 for post 11", 
    "post": "http://localhost:8000/api/posts/11/"
}
```

The next task would be to create the custom Hyperlinked serializer field that generates one URL using the DjangoFilterBackend or using embedded URLs (e.g. `/api/posts/11/comments/`). Is this something that you think is useful? If not, I'll just drop this and use coalesceFindRequests.

dustinfarris commented:
@benkonrath I am working on a comment for this, just want to verify a couple things.

benkonrath commented:
@dustinfarris Great, thanks. In the meantime, I've whipped up a hyperlinked serializer field that generates the url with related query param.

```Python
class DjangoFilterHyperlinkedRelatedField(HyperlinkedRelatedField):

    def __init__(self, view_name=None, **kwargs):
        # TODO: Try to get related_query_param_key automatically from the related view.
        self.related_query_param_key = kwargs.pop('related_query_param_key', None)
        super(DjangoFilterHyperlinkedRelatedField, self).__init__(view_name, **kwargs)

    def get_url(self, related_manager, view_name, request, format):
        """
        Given a related_manager, return the URL with query param that filters the related objects.

        May raise a `NoReverseMatch` if the `view_name` attribute is not
        configured to correctly match the URL conf.
        """
        # Managers without an instance and unsaved instances will not yet have a valid URL.
        if hasattr(related_manager, 'instance') and hasattr(related_manager.instance, 'pk') and related_manager.instance.pk is None:
            return None

        related_query_param_value = getattr(related_manager.instance, self.lookup_field)
        url = self.reverse(view_name, request=request, format=format)
        return '{0}?{1}={2}'.format(url, self.related_query_param_key, related_query_param_value)
```

It would be used on the serializer like this:

```Python
class PostSerializer(serializers.ModelSerializer):
    comments = DjangoFilterHyperlinkedRelatedField(read_only=True, view_name='comment-list', related_query_param_key='blog')
```

With this in place, the post json looks like this:

```json
{
    "id": 11, 
    "title": "post title 11", 
    "body": "post body 11", 
    "comments": "http://localhost:8000/api/comments/?post=11"
}
```

Having gone through all this, I do feel that the `coalesceFindRequests` solution we have in place accomplishes the same thing as this django-filter solution but has the advantage that it's easier to setup and easier to document. I'm going to switch to using the `coalesceFindRequests` solution for my projects. I'm still interested to hear your opinion / thoughts though.

benkonrath commented:
I'd still like to get this functionality into EDA. I don't want to provide support directly for the django-filter based solution that I described in my previous comment because it's too complex. This means that [HyperLinkedRelatedField](http://www.django-rest-framework.org/api-guide/relations/#hyperlinkedrelatedfield) would not work out of the box but the [HyperLinkedIdentityField](http://www.django-rest-framework.org/api-guide/relations/#hyperlinkedidentityfield) would work. `hasMany` fields would still load related records from a URL - it's just not how DRF represents related records by default.

This isn't too much work to finish up and support at least the HyperLinkedIdentityField would be nice. @dustinfarris @holandes22 What do you guys think?

dustinfarris commented:
Yes, would love to get this in as well.

On the `HyperLinkedIdentityField`, would something like this work for hasMany relationships?:

```python
# models.py

class Person(models.Model):
    
    name = models.CharField(max_length=20)


class Animal(models.Model):

    name = models.CharField(max_length=20)
    owner = models.ForeignKey(Person, related_name='pets')


# serializers.py

class PersonSerializer(serializers.ModelSerializer):

    pets = serializers.HyperLinkedIdentityField(view_name='person-pets')

    class Meta:
        model = Person


class AnimalSerializer(serializers.ModelSerializer):
    
    class Meta:
        model = Animal


# viewsets.py

class PersonViewSet(viewsets.ModelViewSet):
    
    queryset = Person.objects.all()
    serializer_class = PersonSerializer
    
    @detail_route()
    def pets(self, request, pk=None):
        person = self.get_object()
        serializer = AnimalSerializer(person.pets.all(), many=True)
        return Response(serializer.data)
```

benkonrath commented:
Yes, works great! I didn't know about the @detail_route feature. I will finish the PR and write some docs based on your example. Thanks.

dustinfarris commented:
What's the status on this?

benkonrath commented:
I should have an updated version that's ready to merge in the next couple of days.

dustinfarris commented:
Great!  Thanks

benkonrath commented:
Ok, it's ready to go. Feedback is appreciated.

dustinfarris commented:
Looks great!  Will review further tomorrow.

dustinfarris commented on addon/serializers/drf.js:
```diff
@@ -11,6 +11,50 @@ import Ember from 'ember';
+      if (relationship.kind === 'hasMany' || relationship.kind === 'belongsTo') {
+        // Matches strings starting with: https://, http://, //, /
+        var payloadRel = hash[payloadRelKey];
+        if (!Ember.isNone(payloadRel.match) && typeof(payloadRel.match) === 'function' && payloadRel.match(/^((https?:)?\/\/|\/)\w/)) {
```
I am getting `TypeError: Cannot read property 'match' of null` here.  I think it is from one of my records that has an empty belongsTo.  I set up a try/catch/debugger and captured this info:

```
relationship.kind = "belongsTo"
payloadRelKey = "myModel"
payloadRel = null
```

I think you just need to account for the null relationship.

benkonrath replied:
Thanks. I'll fix this and add a test.

benkonrath commented:
I've fixed the null relationship bug. Let me know if you want me to squash these commits into one.

dustinfarris commented:
Fantastic work, @benkonrath!

This is literally a drop-in replacement.  I was able to add the detail route, and then change:

```python
pets = serializers.PrimaryKeyRelatedField(queryset=Animal.objects.all(), many=True, required=False)
```

to

```python
pets = serializers.HyperlinkedIdentityField(view_name='person-pets')
```

and everything just worked!  No changes required on the Ember side.  The documentation looks great too.

Yes, please rebase/squash and I'll merge right away.

dustinfarris commented:
Also, we should note somewhere that this is preferred over coalesceFindRequests since there is a limit on the number of records per request due to URL length restrictions.

dustinfarris commented:
For read-only fields, that is.

benkonrath commented:
I've just made these changes:
* Added the note as you suggested.
* Added an example of paginating the nested related results.
* Squashed / reset the commits (except for the mkdocs fix).

benkonrath commented:
@dustinfarris Thanks to you as well - this was definitely a team effort. :)

dustinfarris commented:
:+1:
